File: groots_app/lib/src/service_layer/commands.dart
abstract class Command {}

class LoginCommand extends Command {
  final String email;
  final String password;
  LoginCommand({required this.email, required this.password});
}

class RegisterCommand extends Command {
  final String username;
  final String email;
  final String password;
  RegisterCommand({required this.username, required this.email, required this.password});
}

class LogoutCommand extends Command {}

class LoadLibraryCommand extends Command {}

class AddTrackCommand extends Command {
  final Map<String, dynamic> payload;
  AddTrackCommand(this.payload);
}

class RemoveTrackCommand extends Command {
  final String trackId;
  RemoveTrackCommand(this.trackId);
}

class PinTrackCommand extends Command {
  final String trackId;
  PinTrackCommand(this.trackId);
}

// ── Album commands ─────────────────────────────────────────────────────────

class LoadAlbumsCommand extends Command {}

class LoadGenresCommand extends Command {}

class CreateAlbumCommand extends Command {
  final String title;
  final String artist;
  final int? year;
  final String? genre;
  final String? description;
  final String? recordingFormat;
  CreateAlbumCommand({
    required this.title,
    required this.artist,
    this.year,
    this.genre,
    this.description,
    this.recordingFormat,
  });
}

class UpdateAlbumCommand extends Command {
  final String albumId;
  final String? title;
  final String? artist;
  final int? year;
  final String? genre;
  final String? description;
  final String? recordingFormat;
  UpdateAlbumCommand({
    required this.albumId,
    this.title,
    this.artist,
    this.year,
    this.genre,
    this.description,
    this.recordingFormat,
  });
}

class DeleteAlbumCommand extends Command {
  final String albumId;
  DeleteAlbumCommand(this.albumId);
}

class AssignTrackToAlbumCommand extends Command {
  final String albumId;
  final String trackId;
  final int? trackNumber;
  AssignTrackToAlbumCommand({
    required this.albumId,
    required this.trackId,
    this.trackNumber,
  });
}

class UnassignTrackFromAlbumCommand extends Command {
  final String albumId;
  final String trackId;
  UnassignTrackFromAlbumCommand({required this.albumId, required this.trackId});
}

// ── Playlist commands ──────────────────────────────────────────────────────

class LoadPlaylistsCommand extends Command {}

class CreatePlaylistCommand extends Command {
  final String name;
  CreatePlaylistCommand(this.name);
}

class RenamePlaylistCommand extends Command {
  final String playlistId;
  final String name;
  RenamePlaylistCommand({required this.playlistId, required this.name});
}

class DeletePlaylistCommand extends Command {
  final String playlistId;
  DeletePlaylistCommand(this.playlistId);
}

class AddTrackToPlaylistCommand extends Command {
  final String playlistId;
  final String trackId;
  AddTrackToPlaylistCommand({required this.playlistId, required this.trackId});
}

class RemoveTrackFromPlaylistCommand extends Command {
  final String playlistId;
  final String trackId;
  RemoveTrackFromPlaylistCommand({required this.playlistId, required this.trackId});
}

// ── Admin commands ─────────────────────────────────────────────────────────

class LoadCentralLibraryCommand extends Command {}

class IngestCentralTrackCommand extends Command {
  final String filename;
  final List<int> content;
  final int fileSizeBytes;
  final String mimeType;
  IngestCentralTrackCommand({
    required this.filename,
    required this.content,
    required this.fileSizeBytes,
    required this.mimeType,
  });
}

class SearchAlbumsCommand extends Command {
  final String query;
  SearchAlbumsCommand(this.query);
}

File: backend/src/groots/domain/commands.py
from dataclasses import dataclass


@dataclass
class RegisterUser:
    username: str
    email: str
    password: str
    role_id: str


@dataclass
class LoginUser:
    email: str
    password: str


@dataclass
class AddTrack:
    user_id: str
    cid: str
    title: str
    artist: str
    duration_seconds: int
    file_size_bytes: int
    album: str | None = None
    album_id: str | None = None
    track_number: int | None = None
    year: int | None = None
    genre: str | None = None
    mime_type: str = "audio/mpeg"
    source: str | None = None
    disc_number: int | None = None
    side: str | None = None


@dataclass
class RemoveTrack:
    user_id: str
    track_id: str


@dataclass
class PinTrack:
    user_id: str
    track_id: str
    cid: str


@dataclass
class UploadTrack:
    user_id: str
    filename: str
    content: bytes
    file_size_bytes: int
    mime_type: str
    source: str | None = None
    hint_artist: str | None = None
    hint_title: str | None = None
    hint_album: str | None = None
    hint_year: int | None = None
    hint_track_number: int | None = None


# ── Album commands ────────────────────────────────────────────────────────────


@dataclass
class CreateAlbum:
    title: str
    artist: str
    created_by: str | None = None  # user_id of creator; None for system
    year: int | None = None
    genre: str | None = None
    description: str | None = None
    recording_format: str | None = None


@dataclass
class UpdateAlbum:
    album_id: str
    requesting_user_id: str  # used for permission check (must be creator or admin)
    title: str | None = None
    artist: str | None = None
    year: int | None = None
    genre: str | None = None
    description: str | None = None
    recording_format: str | None = None


@dataclass
class DeleteAlbum:
    album_id: str
    requesting_user_id: str
    is_admin: bool = False


@dataclass
class UploadAlbumCover:
    user_id: str
    album_id: str
    content: bytes
    filename: str
    mime_type: str


@dataclass
class AssignTrackToAlbum:
    user_id: str
    track_id: str
    album_id: str
    track_number: int | None = None
    disc_number: int | None = None
    side: str | None = None


@dataclass
class UnassignTrackFromAlbum:
    user_id: str
    track_id: str


# ── Playlist commands ─────────────────────────────────────────────────────────


@dataclass
class CreatePlaylist:
    user_id: str
    name: str


@dataclass
class RenamePlaylist:
    user_id: str
    playlist_id: str
    name: str


@dataclass
class DeletePlaylist:
    user_id: str
    playlist_id: str


@dataclass
class AddTrackToPlaylist:
    user_id: str
    playlist_id: str
    track_id: str


@dataclass
class RemoveTrackFromPlaylist:
    user_id: str
    playlist_id: str
    track_id: str


# ── Central library commands ──────────────────────────────────────────────────


@dataclass
class IngestCentralTrack:
    """Admin command: add an audio file to the server-managed central library."""

    filename: str
    content: bytes
    file_size_bytes: int
    mime_type: str
